### PR TITLE
Carry forward upstream refresh token on re-authorization

### LINKED ISSUE #5047: Upstream refresh tokens orphaned on re-authorization
## Summary

Upstream refresh tokens are orphaned when a user re-authorizes. Two independent mechanisms cause this: (1) storage is keyed by an ephemeral session ID that is freshly generated on every `/authorize` call, and (2) the callback stores the new IDP response verbatim, overwriting a previously-stored refresh token with the empty string when the provider does not re-issue one.

The second mechanism is most visible with providers that only return `refresh_token` on first consent or when `prompt=consent` is explicitly passed (Google OAuth 2.0 is a prominent example). After any re-authorization, the stored refresh token is effectively lost, and subsequent refresh attempts fail with `ErrNoRefreshToken`.

## Root cause A — session-keyed storage

`pkg/authserver/server/handlers/authorize.go:91` generates a fresh session ID on every `/authorize`:

```go
pending := &storage.PendingAuthorization{
    ...
    SessionID: rand.Text(),   // ← fresh random on every /authorize
    CreatedAt: time.Now(),
}
```

`pkg/authserver/server/handlers/callback.go:142` stores upstream tokens keyed by that session. The session ID is then embedded as the `tsid` claim in the issued JWT (`pkg/authserver/server/session/session.go:120-123`), and downstream read paths look up by `tsid`/session ID (`pkg/auth/token.go:1189-1196`, `pkg/authserver/storage/types.go:222-240`).

`UserID` (the internal ToolHive user subject) is stored as a field on `UpstreamTokens` (`callback.go:138`) but no read path uses it as a lookup key. It is only used for cleanup/metadata.

**Consequence:** any re-authorization mints a new session; the old session's tokens become unreachable through the normal read path.

## Root cause B — callback clobbers prior refresh token

`pkg/authserver/server/handlers/callback.go:131-142` stores `idpTokens.RefreshToken` verbatim, with no preservation of a previously-stored refresh token:

```go
storageTokens := &storage.UpstreamTokens{
    ProviderID:   providerID,
    AccessToken:  idpTokens.AccessToken,
    RefreshToken: idpTokens.RefreshToken,   // empty if provider didn't re-issue
    ...
}
if err := h.storage.StoreUpstreamTokens(ctx, sessionID, providerID, storageTokens); err != nil {
```

`pkg/authserver/refresher.go:73-75` already has the correct pattern for the refresh path (preserve prior refresh token when provider omits a new one). The callback path does not mirror this.

Even if root cause A is fixed (e.g., switching to `UserID`-keyed storage), root cause B will still overwrite a good refresh token with the empty string on every re-auth where the provider does not re-issue a refresh token.

`pkg/auth/upstreamtoken/service.go:135-136` then returns `ErrNoRefreshToken`:

```go
if expired.RefreshToken == "" {
    return nil, ErrNoRefreshToken
}
```

## Failure mode

1. Initial consent completes. Upstream tokens, including a `refresh_token`, are stored under session `S1` with key `(S1, providerID)`.
2. Some time later, a new `/authorize` call is made (any reason — including the compound effect of #5046 causing spurious refresh/reauth cycles).
3. Authorize handler mints a new session `S2`. The new JWT's `tsid` claim points at `S2`.
4. Callback completes. If the provider does not re-issue `refresh_token` (e.g., Google without `prompt=consent` — `pkg/authserver/upstream/oidc.go:367-369` shows `prompt=consent` is optional, not default), the callback stores `{AccessToken: new, RefreshToken: ""}` at `(S2, providerID)`.
5. When the new access token expires, the refresher reads from `(S2, providerID)`, finds `RefreshToken == ""`, and returns `ErrNoRefreshToken`. The provider is then omitted from downstream responses with the log line `omitting provider with unrefreshable expired token` (`pkg/auth/upstreamtoken/service.go:114`).
6. The user is forced to re-authorize. The cycle repeats.

## Proposed fix

Two independent changes — neither alone is sufficient.

1. **Key upstream tokens by `UserID`, not session ID.**
   - Add a `(user_id, provider_id) → tokens` index in both memory and Redis storage backends.
   - Update the downstream read path (`pkg/auth/token.go:1189-1196` and friends) to resolve `tsid → user_id` via the session store, then read upstream tokens by `user_id`.
   - This is net-new index work; `UserID` is currently only metadata.

2. **Preserve prior refresh token in the callback.** Before calling `StoreUpstreamTokens`, load any existing record for `(user_id, provider_id)` and, if `idpTokens.RefreshToken == ""`, carry forward the existing `RefreshToken`. Mirror the pattern already used at `pkg/authserver/refresher.go:73-75`.

## Related

Compounds with #5046. The fake 1h expiry described there triggers the re-authorization cascade that exposes this bug.

## Reproduction

Assumes a Google OAuth client (OIDC) is configured as a ToolHive upstream with the default config (no `prompt=consent` set).

1. Start ToolHive with Google configured as an upstream OIDC provider.
2. As a user, complete the initial `/authorize` flow. Accept Google's consent screen.
3. Verify the stored upstream tokens include a non-empty `RefreshToken`. For the in-memory backend this can be inspected via the authserver's storage; for Redis, read the `upstream_tokens:<session_id>:google` key.
4. Trigger a second `/authorize` for the same Google account, without the initial session being invalidated. This can be done by:
   - Calling the OAuth endpoint again from a fresh client registration, or
   - Signing out of the ToolHive client and signing back in, or
   - Any path that causes the client to restart the authorization flow.
5. Complete the flow. Google returns a new access token but **no** `refresh_token` (this is Google's documented behavior without `prompt=consent`).
6. Inspect the stored tokens under the new session's key: `RefreshToken` is now `""`. The previous session's tokens (with the good refresh token) still exist under the old session key but are unreachable via the new JWT's `tsid`.
7. Wait for the new access token to expire (~1 hour with Google's default).
8. Trigger a downstream request that causes `InProcessService.GetAccessTokens` to run. Observe the log line:

   ```
   omitting provider with unrefreshable expired token provider=google
   ```

9. The Google upstream is now omitted from downstream responses until the user re-authorizes — which will reproduce the same failure.

## Caveats

The claim "Google does not re-issue refresh tokens without `prompt=consent`" reflects Google's documented OAuth 2.0 behavior. The repo only shows `prompt=consent` is optional (`pkg/authserver/upstream/oidc.go:367-369`).

### PR DESCRIPTION
**Pull request by jhrozek**
## Summary

This PR addresses **Root Cause B** of [#5047](https://github.com/stacklok/toolhive/issues/5047): when a user re-authorizes through the auth server, the OAuth callback writes the IdP-supplied `RefreshToken` verbatim into a new `storage.UpstreamTokens` row. Real IdPs commonly **omit** `refresh_token` on a silent re-auth (Google without `prompt=consent`, Microsoft Entra when consent was already granted, GitHub/Okta/Auth0/Keycloak similarly). In every case the previously-issued RT remains valid at the IdP — but today the callback overwrites it with `""`, orphaning the prior good RT and making the next refresh attempt fail. That forces another re-auth, perpetuating the loop.

The fix mirrors the preservation pattern already in place in `upstreamTokenRefresher.RefreshAndStore`: if the IdP returns no RT, look up the most recent prior row for `(userID, providerID)` and carry the prior RT into the new row. A defensive `prior.UpstreamSubject == providerSubject` guard prevents carrying credentials across distinct upstream identities. Storage failures during the lookup are non-fatal — they log a warning and proceed with the empty RT, which is no worse than the current status quo.

- New storage method `UpstreamTokenStorage.GetLatestUpstreamTokensForUser(ctx, userID, providerID)` keyed on user+provider rather than the primary `(sessionID, providerName)` key. Implemented on both Memory and Redis backends; both backends pick the highest `ExpiresAt` so the carry-forward decision is consistent across deployment shapes. Expired rows are intentionally returned (callers want the RT regardless of access-token expiry).
- Callback wiring in `pkg/authserver/server/handlers/callback.go` extracted into `maybeCarryForwardRefreshToken` (gocyclo budget) and called between `storageTokens` construction and `StoreUpstreamTokens`.
- The Redis backend uses the existing `KeyTypeUserUpstream:{userID}` reverse-index set (already maintained atomically by `storeUpstreamTokensScript`), no new Lua script.
- End-to-end integration test against `mockoidc` + a thin RT-stripping reverse proxy (mockoidc unconditionally issues RTs for `authorization_code` and offers no API to suppress them).

**Out of scope, deferred to a follow-up PR**: Root Cause A — the storage rekey from `(sessionID, providerID)` to `(userID, providerID)` plus a one-time Redis migration. Trey's [comment on #5047](https://github.com/stacklok/toolhive/issues/5047#issuecomment-4336470125) describes the split; this PR is the smaller, surgical fix and is valuable on its own because it ends the orphaning loop even before Root Cause A lands.

Closes #5047 partially (Root Cause B; Root Cause A tracked separately).

## Type of change

- [x] Bug fix
- [ ] New feature
- [ ] Refactoring (no behavior change)
- [ ] Dependency update
- [ ] Documentation
- [ ] Other (describe):

## Test plan

- [x] Unit tests (`task test`)
- [ ] E2E tests (`task test-e2e`)
- [x] Linting (`task lint-fix`)
- [ ] Manual testing (describe below)

Specifically verified:
- `go test -race -count=1 ./pkg/authserver/...` — all 13 packages pass.
- `go test -race -count=1 -v -run TestIntegration_ ./pkg/authserver/` — all 23 integration tests pass, including the new `TestIntegration_Callback_PreservesRefreshTokenOnReauth` and the existing `TestIntegration_FullFlow_NonExpiringUpstreamToken[_Redis]` and `TestIntegration_MultiUpstreamChain_MixedExpiryOrderings[_Redis]` from #5092.
- `task test` — pre-existing failure in `pkg/workloads/TestDefaultManager_ListWorkloadsInGroup/workloads_with_empty_group_names` is unrelated to this branch.

## API Compatibility

- [x] This PR does not break the `v1beta1` API, OR the `api-break-allowed` label is applied and the migration guidance is described above.

The new method on `storage.UpstreamTokenStorage` is an internal Go interface, not a CRD field. No CRD schema changes.

## Changes

| File | Change |
|------|--------|
| `pkg/authserver/storage/types.go` | Adds `GetLatestUpstreamTokensForUser` to the `UpstreamTokenStorage` interface with explicit godoc on liveness contract (expired rows are returned) and cross-identity caller responsibility. |
| `pkg/authserver/storage/memory.go` | Implements the new method on `MemoryStorage` — RLock'd iteration, deep-copy on hit, wrapped `ErrNotFound` on miss. |
| `pkg/authserver/storage/redis.go` | Implements the new method on `RedisStorage` using `SMembers` over the existing user-upstream reverse index, single `MGet`, deserialized-`ProviderID` filtering. Tolerates dangling SMEMBERs and the `null` deletion tombstone. |
| `pkg/authserver/storage/mocks/mock_storage.go` | Regenerated. |
| `pkg/authserver/server/handlers/callback.go` | New `maybeCarryForwardRefreshToken` helper invoked between struct construction and `StoreUpstreamTokens`. `UpstreamSubject` guard, non-fatal on storage error. |
| `pkg/authserver/storage/memory_test.go` | Table-driven test for the Memory backend (9 subtests including deep-copy assertion, latest-by-`ExpiresAt`, empty-input validation). |
| `pkg/authserver/storage/redis_test.go` | Table-driven test for the Redis backend (9 subtests including the dangling-SMEMBER stale-index case via miniredis `SAdd`, two-zero-`ExpiresAt` corner case). |
| `pkg/authserver/server/handlers/callback_test.go` | Table-driven `TestCallbackHandler_RefreshTokenCarryForward` consolidating the 5 callback scenarios (preserve, subject mismatch, fresh wins, no prior row, storage error). |
| `pkg/authserver/server/handlers/helpers_test.go` | New `withGetLatestUpstreamTokensError` option mirroring the existing `withStorePendingError` pattern; mock checks it first. |
| `pkg/authserver/integration_test.go` | `TestIntegration_Callback_PreservesRefreshTokenOnReauth` — full callback flow against mockoidc with an RT-stripping reverse proxy. |

## Does this introduce a user-facing change?

Yes. Users who previously hit a refresh-token-orphaning loop on re-authorization (most commonly with Google, Entra, Okta, GitHub, Auth0, Keycloak — all of which omit `refresh_token` on silent re-auth) will see refreshes succeed instead of failing into another forced re-auth. The fix is silent: no configuration changes are required.

## Implementation plan

This PR was planned with Claude Code via a per-step `go-expert-developer` + MoE (`code-reviewer` + `go-security-reviewer` / `oauth-expert` / `redis-valkey-advisor`) workflow. Each step landed as a separate commit with independent review.

<details>
<summary>Approved plan (5 steps)</summary>

1. **Add `GetLatestUpstreamTokensForUser` to storage interface + stub implementations + mock regen.** Keeps the diff tiny so the interface surface is reviewed independently.
2. **Implement Memory backend + 9-case table-driven test.** Filters by `(UserID, ProviderID)`, picks highest `ExpiresAt`, returns deep copy.
3. **Implement Redis backend + 9-case table-driven test.** Uses the existing `KeyTypeUserUpstream` reverse set; SMembers → MGet → unmarshal → filter; tolerates stale index entries and `null` tombstones.
4. **Wire the OAuth callback handler + 5-case table-driven test.** Helper `maybeCarryForwardRefreshToken` between `storageTokens` construction and `StoreUpstreamTokens`. Non-fatal on storage error. `UpstreamSubject` guard.
5. **End-to-end integration test against mockoidc + RT-stripping reverse proxy.** Three legs: initial RT issued → strip flag on, second auth carries forward → third auth as different user gets `ErrNotFound` and stays empty.

Cross-project research (via MCP-mounted source): Hydra delegates upstream-token storage to its consent app and avoids the bug class. oauth2-proxy has the same blind-overwrite shape but sidesteps it via opaque-ticket session keys (orphaned rows TTL out instead of being clobbered) — their `IDToken` preservation pattern at `pkg/providers/oidc.go:140-180` is the explicit code-level analog we mirror for `RefreshToken`. infratographer/identity-api is RFC-8693 only, no broker.

</details>

## Special notes for reviewers

- **This is Root Cause B only.** Root Cause A — the full storage rekey from `(sessionID, providerID)` to `(userID, providerID)` with a one-time Redis migration — is intentionally deferred to a follow-up PR. See Trey's plan in #5047. This fix is valuable on its own: even with Root Cause A unresolved, carrying the RT forward to the new sessionID lets the refresher succeed when the new access token expires, ending the re-auth cycle.
- The `UpstreamSubject == providerSubject` guard is **defense-in-depth**. Given the current identity model (`ResolveUser` keys identity by `(providerID, providerSubject)`), the same internal `UserID` cannot legitimately have two different upstream subjects on the same provider — same-provider account-linking is a future TODO in `storage/types.go`. The guard exists so that future code can't accidentally violate that.
- The integration test does **not** request `offline_access` because mockoidc's `ScopesSupported` rejects it as `invalid_scope`. mockoidc unconditionally issues RTs regardless of scope, so the regression assertion (leg 2 RT matches leg 1 RT) is unaffected — but reviewers should know the test does not exercise the realistic precondition that real IdPs gate RT issuance on `offline_access`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Large PR Justification

- the new integration test is the culprit

### CODE REVIEW

File: pkg/authserver/storage/redis.go [hunk 1]
```diff
@@ -1124,6 +1124,111 @@ func (s *RedisStorage) DeleteUpstreamTokens(ctx context.Context, sessionID strin
...
+	for i, val := range values {
+		stored, ok := parseUserUpstreamEntry(val, providerID, members[i])
+		if !ok {
+			continue
+		}
+
+		// Pick the row with the highest ExpiresAt. Zero ExpiresAt (the no-expiry
+		// sentinel) is represented as 0 — the smallest int64 — so it loses to any
+		// real timestamp unless it is the only candidate.
+		if winner == nil || stored.ExpiresAt > winner.ExpiresAt {
```
Comment (tgrunnagle) [inline pkg/authserver/storage/redis.go]: Question: should we prefer non-expiring tokens?

File: pkg/authserver/storage/redis.go [hunk 2]
```diff
@@ -1124,6 +1124,111 @@ func (s *RedisStorage) DeleteUpstreamTokens(ctx context.Context, sessionID strin
...
+			winnerTokens = &UpstreamTokens{
+				ProviderID:      stored.ProviderID,
+				AccessToken:     stored.AccessToken,
+				RefreshToken:    stored.RefreshToken,
+				IDToken:         stored.IDToken,
+				ExpiresAt:       expiresAt,
+				UserID:          stored.UserID,
+				UpstreamSubject: stored.UpstreamSubject,
+				ClientID:        stored.ClientID,
+			}
```
Comment (tgrunnagle) [inline pkg/authserver/storage/redis.go]: nit: can this happen outside of the loop, referring to `winner` instead of `stored`?

**Review by tgrunnagle — COMMENTED** [review]
## Multi-Agent Consensus Review (MEDIUM findings only)
**Agents consulted**: oauth-flow, security, redis-storage, test-coverage, go-style, general-quality (6 specialists, parallel review)

A full consensus review surfaced 18 actionable findings. Posting only the MEDIUM-severity items here; the LOW findings are polish and didn't feel worth the inline noise on a focused fix PR.

### MEDIUM Findings
| # | Finding | Consensus | Action |
|---|---------|-----------|--------|
| F1 | `SessionExpiresAt` dropped from returned record (Memory + Redis) | 10/10 | Fix — see inline suggestions |
| F7 | Missing handler test — prior row with empty `RefreshToken` | 9/10 | Add table case |

### Overall
The carry-forward design is sound and the implementation is correct for the documented use case. Mirroring `RefreshAndStore`'s preservation pattern at the cross-row level (with the `UpstreamSubject` defense-in-depth guard) is a sensible surgical fix for Root Cause B. Scope is one logical change, the integration-test harness (`rtStrippingProxy`) is genuine novel infrastructure, and the explicit deferral of Root Cause A is appropriate.

The latent bug worth fixing in this PR is **F1**: both backends silently omit `SessionExpiresAt` from the returned struct in `GetLatestUpstreamTokensForUser`. No production caller reads it today, but it's a contract-shaped trap for any future caller that uses this method as a substitute for `GetUpstreamTokens` — exactly the kind of expansion the deferred Root Cause A rekey will encourage. One-line fix per backend; suggestion blocks attached.

**F7** is a test gap: the production guard `prior.RefreshToken != ""` (callback.go:187) is currently untested. The case matters in real chains because the carry-forward chain itself can leave rows with empty RTs, so the guard is load-bearing.

---
*Generated with Claude Code*

File: pkg/authserver/storage/memory.go [hunk 3]
```diff
@@ -846,6 +846,48 @@ func (s *MemoryStorage) DeleteUpstreamTokens(_ context.Context, sessionID string
...
+	if winner == nil {
+		return nil, fmt.Errorf("%w: %w", ErrNotFound, fosite.ErrNotFound.WithHint("Upstream tokens not found"))
+	}
+
+	return &UpstreamTokens{
+		ProviderID:      winner.ProviderID,
+		AccessToken:     winner.AccessToken,
+		RefreshToken:    winner.RefreshToken,
+		IDToken:         winner.IDToken,
+		ExpiresAt:       winner.ExpiresAt,
```
Comment (tgrunnagle) [inline pkg/authserver/storage/memory.go]: **F1 — `SessionExpiresAt` dropped from returned record** (10/10, MEDIUM)

The sibling `GetUpstreamTokens` (memory.go:773-783) populates `SessionExpiresAt` but this new method omits it. Today's only caller (`maybeCarryForwardRefreshToken`) reads only `RefreshToken` and `UpstreamSubject`, so the bug is latent — but the interface returns `*UpstreamTokens` with no "partial copy" caveat in the godoc, so any future caller (e.g. the deferred Root Cause A rekey) will get a silently-zero `SessionExpiresAt` and break the storage TTL invariant the refresher relies on (refresher.go:68-85).

Raised by: redis-storage, go-style, general-quality

```suggestion
		ExpiresAt:        winner.ExpiresAt,
		SessionExpiresAt: winner.SessionExpiresAt,
```

File: pkg/authserver/storage/redis.go [hunk 4]
```diff
@@ -1124,6 +1124,111 @@ func (s *RedisStorage) DeleteUpstreamTokens(ctx context.Context, sessionID strin
...
+			var expiresAt time.Time
+			if stored.ExpiresAt != 0 {
+				expiresAt = time.Unix(stored.ExpiresAt, 0)
+			}
+			winnerTokens = &UpstreamTokens{
+				ProviderID:      stored.ProviderID,
+				AccessToken:     stored.AccessToken,
+				RefreshToken:    stored.RefreshToken,
+				IDToken:         stored.IDToken,
+				ExpiresAt:       expiresAt,
```
Comment (tgrunnagle) [inline pkg/authserver/storage/redis.go]: **F1 — `SessionExpiresAt` dropped from returned record** (10/10, MEDIUM)

Same issue as the Memory backend: `unmarshalUpstreamTokens` (redis.go:1163-1166) populates `SessionExpiresAt`, this method does not. Field-list duplication here means there are now four copies of the same struct construction in this package — extracting a `(s *storedUpstreamTokens).toUpstreamTokens()` helper would prevent further drift.

```suggestion
				ExpiresAt:       expiresAt,
				SessionExpiresAt: func() time.Time {
					if stored.SessionExpiresAt != 0 {
						return time.Unix(stored.SessionExpiresAt, 0)
					}
					return time.Time{}
				}(),
```

File: pkg/authserver/server/handlers/callback_test.go [hunk 5]
```diff
@@ -767,6 +767,148 @@ func TestCallbackHandler_TwoUpstreams_StorePendingError_CleansUp(t *testing.T) {
...
+func TestCallbackHandler_RefreshTokenCarryForward(t *testing.T) {
+	t.Parallel()
+
+	type priorRow struct {
+		sessionID       string
+		upstreamSubject string
+		refreshToken    string
+	}
+
+	cases := []struct {
```
Comment (tgrunnagle) [inline pkg/authserver/server/handlers/callback_test.go:843]: **F7 — Missing test case: prior row with empty `RefreshToken`** (9/10, MEDIUM)

The production guard `prior.RefreshToken != ""` (callback.go:187) is currently untested. The case matters in real chains: the carry-forward chain itself can leave rows with empty RTs (when leg N had no prior to carry from), so this guard is load-bearing. Suggest adding a table case:

```
{
    name:             "prior row with empty RT does not carry",
    priorRow:         &priorRow{sessionID: "old-session", upstreamSubject: "user-123", refreshToken: ""},
    idpRefreshToken:  "",
    idpSubject:       "user-123",
    expectedStoredRT: "",
},
```

Raised by: test-coverage, security

File: pkg/authserver/storage/memory.go (hunk 3 above)
Comment (jhrozek) [reply pkg/authserver/storage/memory.go]: Fixed in c162013d. Also extracted a `cloneUpstreamTokens` helper in 6d1d2678 since this was the third site doing the same field-by-field copy and the next field addition would hit the same trap. Whole-struct round-trip test in 15391fd9 — a future drop fails loudly.

File: pkg/authserver/storage/redis.go (hunk 4 above)
Comment (jhrozek) [reply pkg/authserver/storage/redis.go]: Fixed in c162013d. Extracted `(*storedUpstreamTokens).toUpstreamTokens()` in 6d1d2678 — kept it as a method on the stored type since Redis converts int64 epoch sentinels to time.Time, while Memory's `cloneUpstreamTokens` is a free function (different semantics). Round-trip test added there too.

File: pkg/authserver/server/handlers/callback_test.go (hunk 5 above)
Comment (jhrozek) [reply pkg/authserver/server/handlers/callback_test.go:843]: Added in 678e415a. While there, broadened the shared assertion block (AccessToken, IDToken, UpstreamSubject, non-zero ExpiresAt across all cases) — without those, a regression that early-returns from the callback before StoreUpstreamTokens would still pass an RT-only assertion.

File: pkg/authserver/storage/redis.go (hunk 1 above)
Comment (jhrozek) [reply pkg/authserver/storage/redis.go]: Good catch — flipped in f564b0f8. After re-reading #5092, the old rule was inconsistent with the rest of the package (IsExpired, cleanupExpired, marshalUpstreamTokensWithTTL all treat zero ExpiresAt as "alive forever" for Slack and GitHub OAuth Apps). Carry-forward caller is largely unaffected today since those rows have empty RefreshToken and the guard rejects them anyway, but the flipped rule is the more sensible default for any future caller. Note that #5047 Scope A's rekey to (userID, providerID) collapses this to one row per pair and makes the whole tie-breaker vestigial — so this is also a stopgap on that road.

File: pkg/authserver/storage/redis.go (hunk 2 above)
Comment (jhrozek) [reply pkg/authserver/storage/redis.go]: Done in 6d1d2678. Loop now tracks `winner *storedUpstreamTokens` and calls `winner.toUpstreamTokens()` once after the loop.

**Comment by jhrozek** [comment]
Thanks for the review. Both MEDIUM findings and both nits addressed — replied on each thread with specifics.

**Review by tgrunnagle — APPROVED** [review]
